# Problem Statement

To create a regression model that predicts salaries of baseball players based on their statistics.

# Importing Libraries

In [52]:
import warnings
warnings.simplefilter(action='ignore')
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt


from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.linear_model import LinearRegression, Ridge, RidgeCV, ElasticNet, Lasso, LassoCV
from sklearn.metrics import mean_squared_error
from sklearn.neighbors import LocalOutlierFactor
from sklearn.preprocessing import LabelEncoder
from sklearn import preprocessing
from sklearn.preprocessing import RobustScaler

In [53]:
Hitters=pd.read_csv("C:\\Users\\Rushabh Shah\\Downloads\\Hitters.csv")

# Loading & understanding of dataset

In [54]:
Hitters.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 322 entries, 0 to 321
Data columns (total 20 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   AtBat      322 non-null    int64  
 1   Hits       322 non-null    int64  
 2   HmRun      322 non-null    int64  
 3   Runs       322 non-null    int64  
 4   RBI        322 non-null    int64  
 5   Walks      322 non-null    int64  
 6   Years      322 non-null    int64  
 7   CAtBat     322 non-null    int64  
 8   CHits      322 non-null    int64  
 9   CHmRun     322 non-null    int64  
 10  CRuns      322 non-null    int64  
 11  CRBI       322 non-null    int64  
 12  CWalks     322 non-null    int64  
 13  League     322 non-null    object 
 14  Division   322 non-null    object 
 15  PutOuts    322 non-null    int64  
 16  Assists    322 non-null    int64  
 17  Errors     322 non-null    int64  
 18  Salary     263 non-null    float64
 19  NewLeague  322 non-null    object 
dtypes: float64

In [55]:
Hitters.describe().T

count         mean          std   min     25%     50%      75%  \
AtBat    322.0   380.928571   153.404981  16.0  255.25   379.5   512.00   
Hits     322.0   101.024845    46.454741   1.0   64.00    96.0   137.00   
HmRun    322.0    10.770186     8.709037   0.0    4.00     8.0    16.00   
Runs     322.0    50.909938    26.024095   0.0   30.25    48.0    69.00   
RBI      322.0    48.027950    26.166895   0.0   28.00    44.0    64.75   
Walks    322.0    38.742236    21.639327   0.0   22.00    35.0    53.00   
Years    322.0     7.444099     4.926087   1.0    4.00     6.0    11.00   
CAtBat   322.0  2648.683230  2324.205870  19.0  816.75  1928.0  3924.25   
CHits    322.0   717.571429   654.472627   4.0  209.00   508.0  1059.25   
CHmRun   322.0    69.490683    86.266061   0.0   14.00    37.5    90.00   
CRuns    322.0   358.795031   334.105886   1.0  100.25   247.0   526.25   
CRBI     322.0   330.118012   333.219617   0.0   88.75   220.5   426.25   
CWalks   322.0   260.239130   267.058085   0.0   67.25   170.5   339.25   
PutOuts  322.0   288.937888   280.704614   0.0  109.25   212.0   325.00   
Assists  322.0   106.913043   136.854876   0.0    7.00    39.5   166.00   
Errors   322.0     8.040373     6.368359   0.0    3.00     6.0    11.00   
Salary   263.0   535.925882   451.118681  67.5  190.00   425.0   750.00   

             max  
AtBat      687.0  
Hits       238.0  
HmRun       40.0  
Runs       130.0  
RBI        121.0  
Walks      105.0  
Years       24.0  
CAtBat   14053.0  
CHits     4256.0  
CHmRun     548.0  
CRuns     2165.0  
CRBI      1659.0  
CWalks    1566.0  
PutOuts   1378.0  
Assists    492.0  
Errors      32.0  
Salary    2460.0

In [56]:
Hitters.isnull().sum()

AtBat         0
Hits          0
HmRun         0
Runs          0
RBI           0
Walks         0
Years         0
CAtBat        0
CHits         0
CHmRun        0
CRuns         0
CRBI          0
CWalks        0
League        0
Division      0
PutOuts       0
Assists       0
Errors        0
Salary       59
NewLeague     0
dtype: int64

In [57]:
corr_matrix=Hitters.corr().abs()
Hitters.corr()

AtBat      Hits     HmRun      Runs       RBI     Walks     Years  \
AtBat    1.000000  0.967939  0.592198  0.913060  0.820539  0.669845  0.047372   
Hits     0.967939  1.000000  0.562158  0.922187  0.811073  0.641211  0.044767   
HmRun    0.592198  0.562158  1.000000  0.650988  0.855122  0.481014  0.116318   
Runs     0.913060  0.922187  0.650988  1.000000  0.798206  0.732213  0.004541   
RBI      0.820539  0.811073  0.855122  0.798206  1.000000  0.615997  0.146168   
Walks    0.669845  0.641211  0.481014  0.732213  0.615997  1.000000  0.136475   
Years    0.047372  0.044767  0.116318  0.004541  0.146168  0.136475  1.000000   
CAtBat   0.235526  0.227565  0.221882  0.186497  0.294688  0.277175  0.920289   
CHits    0.252717  0.255815  0.220627  0.204830  0.308201  0.280671  0.903631   
CHmRun   0.236659  0.202712  0.493227  0.227913  0.441771  0.332473  0.726872   
CRuns    0.266534  0.261787  0.262361  0.250556  0.323285  0.338478  0.882877   
CRBI     0.244053  0.232005  0.351979  0.205976  0.393184  0.308631  0.868812   
CWalks   0.166123  0.151818  0.233154  0.182168  0.250914  0.424507  0.838533   
PutOuts  0.317550  0.310673  0.282923  0.279347  0.343186  0.299515 -0.004684   
Assists  0.353824  0.320455 -0.106329  0.220567  0.106591  0.149656 -0.080638   
Errors   0.352117  0.310038  0.039318  0.240475  0.193370  0.129382 -0.162140   
Salary   0.394771  0.438675  0.343028  0.419859  0.449457  0.443867  0.400657   

           CAtBat     CHits    CHmRun     CRuns      CRBI    CWalks   PutOuts  \
AtBat    0.235526  0.252717  0.236659  0.266534  0.244053  0.166123  0.317550   
Hits     0.227565  0.255815  0.202712  0.261787  0.232005  0.151818  0.310673   
HmRun    0.221882  0.220627  0.493227  0.262361  0.351979  0.233154  0.282923   
Runs     0.186497  0.204830  0.227913  0.250556  0.205976  0.182168  0.279347   
RBI      0.294688  0.308201  0.441771  0.323285  0.393184  0.250914  0.343186   
Walks    0.277175  0.280671  0.332473  0.338478  0.308631  0.424507  0.299515   
Years    0.920289  0.903631  0.726872  0.882877  0.868812  0.838533 -0.004684   
CAtBat   1.000000  0.995063  0.798836  0.983345  0.949219  0.906501  0.062283   
CHits    0.995063  1.000000  0.783306  0.984609  0.945141  0.890954  0.076547   
CHmRun   0.798836  0.783306  1.000000  0.820243  0.929484  0.799983  0.112724   
CRuns    0.983345  0.984609  0.820243  1.000000  0.943769  0.927807  0.064180   
CRBI     0.949219  0.945141  0.929484  0.943769  1.000000  0.884726  0.110098   
CWalks   0.906501  0.890954  0.799983  0.927807  0.884726  1.000000  0.058638   
PutOuts  0.062283  0.076547  0.112724  0.064180  0.110098  0.058638  1.000000   
Assists  0.002038 -0.002523 -0.158511 -0.022978 -0.079387 -0.039130 -0.025024   
Errors  -0.066922 -0.062756 -0.138115 -0.084395 -0.100990 -0.118475  0.109972   
Salary   0.526135  0.548910  0.524931  0.562678  0.566966  0.489822  0.300480   

          Assists    Errors    Salary  
AtBat    0.353824  0.352117  0.394771  
Hits     0.320455  0.310038  0.438675  
HmRun   -0.106329  0.039318  0.343028  
Runs     0.220567  0.240475  0.419859  
RBI      0.106591  0.193370  0.449457  
Walks    0.149656  0.129382  0.443867  
Years   -0.080638 -0.162140  0.400657  
CAtBat   0.002038 -0.066922  0.526135  
CHits   -0.002523 -0.062756  0.548910  
CHmRun  -0.158511 -0.138115  0.524931  
CRuns   -0.022978 -0.084395  0.562678  
CRBI    -0.079387 -0.100990  0.566966  
CWalks  -0.039130 -0.118475  0.489822  
PutOuts -0.025024  0.109972  0.300480  
Assists  1.000000  0.706362  0.025436  
Errors   0.706362  1.000000 -0.005401  
Salary   0.025436 -0.005401  1.000000

In [58]:
Hitters['Year_lab'] = pd.cut(x=Hitters['Years'], bins=[0, 3, 6, 10, 15, 19, 24])
Hitters.groupby(['League','Division', 'Year_lab']).agg({'Salary':'mean'})

Salary
League Division Year_lab            
A      E        (0, 3]    112.500000
                (3, 6]    655.568182
                (6, 10]   852.738125
                (10, 15]  816.311353
                (15, 19]  665.416750
                (19, 24]         NaN
       W        (0, 3]    153.613636
                (3, 6]    401.360000
                (6, 10]   633.958375
                (10, 15]  835.250000
                (15, 19]  479.000000
                (19, 24]  487.500000
N      E        (0, 3]    248.520813
                (3, 6]    501.191650
                (6, 10]   824.226143
                (10, 15]  894.322667
                (15, 19]  662.500000
                (19, 24]         NaN
       W        (0, 3]    191.766667
                (3, 6]    458.333333
                (6, 10]   563.229187
                (10, 15]  721.894000
                (15, 19]  760.833250
                (19, 24]  475.000000

In [59]:
Hitters['Salary'] = Hitters.groupby(['League', 'Division', 'Year_lab'])['Salary'].transform(lambda x: x.fillna(x.mean()))

#### NA values were filled by looking at "Salary" averages in age, league and division variables, Dummy variables were created. The X variables were normalized.

In [60]:
Hitters.isnull().sum()

AtBat        0
Hits         0
HmRun        0
Runs         0
RBI          0
Walks        0
Years        0
CAtBat       0
CHits        0
CHmRun       0
CRuns        0
CRBI         0
CWalks       0
League       0
Division     0
PutOuts      0
Assists      0
Errors       0
Salary       0
NewLeague    0
Year_lab     0
dtype: int64

In [61]:
Hitters.shape

(322, 21)

In [62]:
Hitters.head()

AtBat  Hits  HmRun  Runs  RBI  Walks  Years  CAtBat  CHits  CHmRun  ...  \
0    293    66      1    30   29     14      1     293     66       1  ...   
1    315    81      7    24   38     39     14    3449    835      69  ...   
2    479   130     18    66   72     76      3    1624    457      63  ...   
3    496   141     20    65   78     37     11    5628   1575     225  ...   
4    321    87     10    39   42     30      2     396    101      12  ...   

   CRBI  CWalks  League Division PutOuts  Assists  Errors  Salary  NewLeague  \
0    29      14       A        E     446       33      20   112.5          A   
1   414     375       N        W     632       43      10   475.0          N   
2   266     263       A        W     880       82      14   480.0          A   
3   838     354       N        E     200       11       3   500.0          N   
4    46      33       N        E     805       40       4    91.5          N   

   Year_lab  
0    (0, 3]  
1  (10, 15]  
2    (0, 3]  
3  (10, 15]  
4    (0, 3]  

[5 rows x 21 columns]

In [63]:
le = LabelEncoder()
Hitters['League'] = le.fit_transform(Hitters['League'])
Hitters['Division'] = le.fit_transform(Hitters['Division'])
Hitters['NewLeague'] = le.fit_transform(Hitters['NewLeague'])
Hitters['Year_lab'] = le.fit_transform(Hitters['Year_lab'])

In [64]:
Hitters.head(50)

AtBat  Hits  HmRun  Runs  RBI  Walks  Years  CAtBat  CHits  CHmRun  ...  \
0     293    66      1    30   29     14      1     293     66       1  ...   
1     315    81      7    24   38     39     14    3449    835      69  ...   
2     479   130     18    66   72     76      3    1624    457      63  ...   
3     496   141     20    65   78     37     11    5628   1575     225  ...   
4     321    87     10    39   42     30      2     396    101      12  ...   
5     594   169      4    74   51     35     11    4408   1133      19  ...   
6     185    37      1    23    8     21      2     214     42       1  ...   
7     298    73      0    24   24      7      3     509    108       0  ...   
8     323    81      6    26   32      8      2     341     86       6  ...   
9     401    92     17    49   66     65     13    5206   1332     253  ...   
10    574   159     21   107   75     59     10    4631   1300      90  ...   
11    202    53      4    31   26     27      9    1876    467      15  ...   
12    418   113     13    48   61     47      4    1512    392      41  ...   
13    239    60      0    30   11     22      6    1941    510       4  ...   
14    196    43      7    29   27     30     13    3231    825      36  ...   
15    183    39      3    20   15     11      3     201     42       3  ...   
16    568   158     20    89   75     73     15    8068   2273     177  ...   
17    190    46      2    24    8     15      5     479    102       5  ...   
18    407   104      6    57   43     65     12    5233   1478     100  ...   
19    127    32      8    16   22     14      8     727    180      24  ...   
20    413    92     16    72   48     65      1     413     92      16  ...   
21    426   109      3    55   43     62      1     426    109       3  ...   
22     22    10      1     4    2      1      6      84     26       2  ...   
23    472   116     16    60   62     74      6    1924    489      67  ...   
24    629   168     18    73  102     40     18    8424   2464     164  ...   
25    587   163      4    92   51     70      6    2695    747      17  ...   
26    324    73      4    32   18     22      7    1931    491      13  ...   
27    474   129     10    50   56     40     10    2331    604      61  ...   
28    550   152      6    92   37     81      5    2308    633      32  ...   
29    513   137     20    90   95     90     14    5201   1382     166  ...   
30    313    84      9    42   30     39     17    6890   1833     224  ...   
31    419   108      6    55   36     22      3     591    149       8  ...   
32    517   141     27    70   87     52      9    3571    994     215  ...   
33    583   168     17    83   80     56      5    1646    452      44  ...   
34    204    49      6    23   25     12      7    1309    308      27  ...   
35    379   106     10    38   60     30     14    6207   1906     146  ...   
36    161    36      0    19   10     17      4    1053    244       3  ...   
37    268    60      5    24   25     15      2     350     78       5  ...   
38    346    98      5    31   53     30     16    5913   1615     235  ...   
39    241    61      1    34   12     14      1     241     61       1  ...   
40    181    41      1    15   21     33      2     232     50       4  ...   
41    216    54      0    21   18     15     18    7318   1926      46  ...   
42    200    57      6    23   14     14      9    2516    684      46  ...   
43    217    46      7    32   19      9      4     694    160      32  ...   
44    194    40      7    19   29     30     11    4183   1069      64  ...   
45    254    68      2    28   26     22      6     999    236      21  ...   
46    416   132      7    57   49     33      3     932    273      24  ...   
47    205    57      8    34   32      9      5     756    192      32  ...   
48    542   140     12    46   75     41     16    7099   2130     235  ...   
49    526   146     13    71   70     84      6    264

In [65]:
Hitters.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 322 entries, 0 to 321
Data columns (total 21 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   AtBat      322 non-null    int64  
 1   Hits       322 non-null    int64  
 2   HmRun      322 non-null    int64  
 3   Runs       322 non-null    int64  
 4   RBI        322 non-null    int64  
 5   Walks      322 non-null    int64  
 6   Years      322 non-null    int64  
 7   CAtBat     322 non-null    int64  
 8   CHits      322 non-null    int64  
 9   CHmRun     322 non-null    int64  
 10  CRuns      322 non-null    int64  
 11  CRBI       322 non-null    int64  
 12  CWalks     322 non-null    int64  
 13  League     322 non-null    int32  
 14  Division   322 non-null    int32  
 15  PutOuts    322 non-null    int64  
 16  Assists    322 non-null    int64  
 17  Errors     322 non-null    int64  
 18  Salary     322 non-null    float64
 19  NewLeague  322 non-null    int32  
 20  Year_lab  

# Normalization

In [66]:
Hitters_X= Hitters.drop(["Salary","League","Division","NewLeague"], axis=1)

scaled_cols5=preprocessing.normalize(Hitters_X)

scaled_cols=pd.DataFrame(scaled_cols5, columns=Hitters_X.columns)
scaled_cols.head()

AtBat      Hits     HmRun      Runs       RBI     Walks     Years  \
0  0.472401  0.106411  0.001612  0.048369  0.046756  0.022572  0.001612   
1  0.085657  0.022026  0.001903  0.006526  0.010333  0.010605  0.003807   
2  0.237036  0.064331  0.008907  0.032660  0.035630  0.037609  0.001485   
3  0.082624  0.023488  0.003332  0.010828  0.012993  0.006163  0.001832   
4  0.331579  0.089867  0.010330  0.040285  0.043384  0.030989  0.002066   

     CAtBat     CHits    CHmRun     CRuns      CRBI    CWalks   PutOuts  \
0  0.472401  0.106411  0.001612  0.048369  0.046756  0.022572  0.719082   
1  0.937879  0.227060  0.018763  0.087289  0.112578  0.101973  0.171858   
2  0.803645  0.226149  0.031176  0.110848  0.131631  0.130147  0.435473   
3  0.937518  0.262365  0.037481  0.137929  0.139595  0.058970  0.033316   
4  0.409050  0.104328  0.012395  0.049582  0.047516  0.034088  0.831529   

    Assists    Errors  Year_lab  
0  0.053206  0.032246  0.000000  
1  0.011693  0.002719  0.000816  
2  0.040578  0.006928  0.000000  
3  0.001832  0.000500  0.000500  
4  0.041318  0.004132  0.000000

In [67]:
cat_Hitters=pd.concat([Hitters.loc[:,"League":"Division"],Hitters.loc[:,"NewLeague":"Year_lab"]], axis=1)
cat_Hitters.head()

League  Division  NewLeague  Year_lab
0       0         0          0         0
1       1         1          1         3
2       0         1          0         0
3       1         0          1         3
4       1         0          1         0

In [68]:
Hitters= pd.concat([scaled_cols,cat_Hitters,Hitters["Salary"]], axis=1)
Hitters

AtBat      Hits     HmRun      Runs       RBI     Walks     Years  \
0    0.472401  0.106411  0.001612  0.048369  0.046756  0.022572  0.001612   
1    0.085657  0.022026  0.001903  0.006526  0.010333  0.010605  0.003807   
2    0.237036  0.064331  0.008907  0.032660  0.035630  0.037609  0.001485   
3    0.082624  0.023488  0.003332  0.010828  0.012993  0.006163  0.001832   
4    0.331579  0.089867  0.010330  0.040285  0.043384  0.030989  0.002066   
..        ...       ...       ...       ...       ...       ...       ...   
317  0.169544  0.043324  0.002388  0.022174  0.016374  0.012622  0.001706   
318  0.083222  0.023004  0.000846  0.012855  0.008457  0.015900  0.002030   
319  0.256903  0.068147  0.001623  0.032992  0.023256  0.028124  0.003245   
320  0.155442  0.039064  0.002441  0.023059  0.016277  0.021160  0.002170   
321  0.120098  0.032356  0.001713  0.014655  0.008375  0.005900  0.002094   

       CAtBat     CHits    CHmRun  ...    CWalks   PutOuts   Assists  \
0    0.472401  0.106411  0.001612  ...  0.022572  0.719082  0.053206   
1    0.937879  0.227060  0.018763  ...  0.101973  0.171858  0.011693   
2    0.803645  0.226149  0.031176  ...  0.130147  0.435473  0.040578   
3    0.937518  0.262365  0.037481  ...  0.058970  0.033316  0.001832   
4    0.409050  0.104328  0.012395  ...  0.034088  0.831529  0.041318   
..        ...       ...       ...  ...       ...       ...       ...   
317  0.922085  0.274954  0.010916  ...  0.047076  0.110869  0.003070   
318  0.932185  0.255585  0.006597  ...  0.148006  0.052944  0.064446   
319  0.919443  0.234188  0.003786  ...  0.078964  0.020011  0.061116   
320  0.867543  0.232484  0.026314  ...  0.090064  0.356458  0.035537   
321  0.934139  0.277311  0.005710  ...  0.047392  0.077655  0.000761   

       Errors  Year_lab  League  Division  NewLeague  Year_lab  Salary  
0    0.032246  0.000000       0         0          0         0   112.5  
1    0.002719  0.000816       1         1          1         3   475.0  
2    0.006928  0.000000       0         1          0         0   480.0  
3    0.000500  0.000500       1         0          1         3   500.0  
4    0.004132  0.000000       1         0          1         0    91.5  
..        ...       ...     ...       ...        ...       ...     ...  
317  0.001023  0.000341       1         0          1         1   700.0  
318  0.003383  0.000507       0         0          0         3   875.0  
319  0.003786  0.000541       0         1          0         1   385.0  
320  0.003255  0.000543       0         0          0         2   960.0  
321  0.000571  0.000571       0         1          0         3  1000.0  

[322 rows x 22 columns]

# Modelling

### Ridge Regression & Model tuning

In [69]:
y=Hitters["Salary"]
X=Hitters.drop("Salary", axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y,test_size=0.20, random_state=46)

ridreg = Ridge()
model = ridreg.fit(X_train, y_train)
y_pred = model.predict(X_test)
Hitters_ridreg_rmse = np.sqrt(mean_squared_error(y_test,y_pred))
Hitters_ridreg_rmse 

313.57725335744357

In [70]:
y=Hitters["Salary"]
X=Hitters.drop("Salary", axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=46)


alpha = [1e-20,1e-15, 1e-10, 1e-8, 1e-5,1e-4, 1e-3,1e-2,1,2,5,10,20,50,100,200]
ridreg_cv = RidgeCV(alphas = alpha, scoring = "neg_mean_squared_error", cv =10, normalize = True)
ridreg_cv.fit(X_train, y_train)
print(ridreg_cv.alpha_)

#Final Model

ridreg_tuned = Ridge(alpha = ridreg_cv.alpha_).fit(X_train,y_train)
y_pred = ridreg_tuned.predict(X_test)
Hitters_ridge_tuned_rmse = np.sqrt(mean_squared_error(y_test,y_pred))
Hitters_ridge_tuned_rmse

0.001


294.5139057976824

### Lasso Regression & Model Tuning

In [71]:
y=Hitters["Salary"]
X=Hitters.drop("Salary", axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=46)

lasreg = Lasso()
model = lasreg.fit(X_train,y_train)
y_pred = model.predict(X_test)
Hitters_lasreg_rmse = np.sqrt(mean_squared_error(y_test,y_pred))
Hitters_lasreg_rmse

287.03573553221474

In [79]:
y=Hitters["Salary"]
X=Hitters.drop("Salary", axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=46)

alpha = [0.1,0.2,0.3,0.5,0.8,0.9,1]
lasso_cv = LassoCV(alphas = alpha, cv = 10, normalize = True)
lasso_cv.fit(X_train, y_train)
print(lasso_cv.alpha_)

#Final Model

lasso_tuned = Lasso(alpha = lasso_cv.alpha_).fit(X_train,y_train)
y_pred = lasso_tuned.predict(X_test)
Hitters_lasso_tuned_rmse = np.sqrt(mean_squared_error(y_test, y_pred))
Hitters_lasso_tuned_rmse

0.1


283.4926545111205

# Final Output & Conclusion

In [80]:
Regression_result_hitters = pd.DataFrame({"CONDITIONS":["Hitters: filled with mean",],
                              "RIDGE":[Hitters_ridreg_rmse],
                              "RIDGE TUNED":[Hitters_ridge_tuned_rmse],
                              "LASSO":[Hitters_lasreg_rmse],
                              "LASSO TUNED":[Hitters_lasso_tuned_rmse],})

Regression_result_hitters

CONDITIONS       RIDGE  RIDGE TUNED       LASSO  LASSO TUNED
0  Hitters: filled with mean  313.577253   294.513906  287.035736   283.492655

#### We have applied regression model with ridge and lasso method. And The lowets error is in Lasso Tuned Model Model with RMSE value of  283.49 with and alpha value of 0.1.